File: zZ/java/valid-parentheses.java
/**
 * Valid Parentheses
 * 
 * Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
 * 
 * The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
 */

public class Solution {
    public boolean isValid(String s) {
        if (null == s) {
            throw new NullPointerException();
        }

        Stack<Character> stackForLeftParentheses = new Stack<Character>();
        for (int i = 0; i < s.length(); ++i) {
            char character = s.charAt(i);
            if (character == '(' ||
                character == '[' ||
                character == '{') {
                stackForLeftParentheses.push(character);
            } else {
                if (stackForLeftParentheses.isEmpty()) {
                    return false;
                }

                switch (character) {
                    case ')':
                        if (stackForLeftParentheses.peek() != '(') {
                            return false;
                        } else {
                            stackForLeftParentheses.pop();
                        }
                        break;
                    case ']':
                        if (stackForLeftParentheses.peek() != '[') {
                            return false;
                        } else {
                            stackForLeftParentheses.pop();
                        }
                        break;
                    case '}':
                        if (stackForLeftParentheses.peek() != '{') {
                            return false;
                        } else {
                            stackForLeftParentheses.pop();
                        }
                        break;
                }
            }
        }

        return stackForLeftParentheses.isEmpty();
    }
}

File: data-structure-and-algorithm/typescript/leetcode/src/valid-parentheses.ts
/*
Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.


Example 1:

Input: s = "()"
Output: true
Example 2:

Input: s = "()[]{}"
Output: true
Example 3:

Input: s = "(]"
Output: false
Example 4:

Input: s = "([)]"
Output: false
Example 5:

Input: s = "{[]}"
Output: true


Constraints:

1 <= s.length <= 104
s consists of parentheses only '()[]{}'.
*/

import {Stack} from './data-structure/Stack';

function isValidParenthesis(input: string): boolean {
  const left = '([{';
  const parenthesisMap = new Map<string, string>([
    [')', '('],
    ['}', '{'],
    [']', '['],
  ]);
  const stack = new Stack<string>();
  for (const char of input) {
    if (left.includes(char)) {
      stack.push(char);
    } else if (parenthesisMap.get(char) !== stack.peek()) {
      return false;
    } else {
      stack.pop();
    }
  }

  return stack.isEmpty();
}
